crates/milli/src/progress.rs

```diff
@@ -50,18 +50,58 @@ struct InnerProgress {
 }
 
 impl Progress {
-    pub fn update_progress<P: Step>(&self, sub_progress: P) {
+    /// Update the progress and return `true` if the step was started, `false` if it was already started.
+    pub fn update_progress<P: Step>(&self, sub_progress: P) -> bool {
         let mut inner = self.steps.write().unwrap();
         let InnerProgress { steps, durations } = &mut *inner;
 
         let now = Instant::now();
         let step_type = TypeId::of::<P>();
         if let Some(idx) = steps.iter().position(|(id, _, _)| *id == step_type) {
+            if steps[idx].1.name() == sub_progress.name() {
+                // The step is already started, so we don't need to start it again.
+                return false;
+            }
+
             push_steps_durations(steps, durations, now, idx);
             steps.truncate(idx);
         }
 
         steps.push((step_type, Box::new(sub_progress), now));
+        true
+    }
+
+    /// End a step that has been started without having to start a new step.
+    fn end_progress_step<P: Step>(&self, sub_progress: P) {
+        let mut inner = self.steps.write().unwrap();
+        let InnerProgress { steps, durations } = &mut *inner;
+
+        let now = Instant::now();
+        let step_type = TypeId::of::<P>();
+
+        debug_assert!(
+            steps.iter().any(|(id, s, _)| *id == step_type && s.name() == sub_progress.name()),
+            "Step `{}` must have been started",
+            sub_progress.name()
+        );
```

Comment:
if this debug assert fires, then the lock will be poisoned.

I'm also not sure about having asserts that only fire in debug mode.

If we want to keep it, I say we should have a cfg(dev) to compute the boolean and then explicitly drop the lock before panicking when the assertion doesn't hold